# Segmenting and Clustering Assignment

## First part 

In [71]:
table = """"

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4B
</td>
<td>East York
</td>
<td>Parkview Hill, Woodbine Gardens
</td></tr>
<tr>
<td>M5B
</td>
<td>Downtown Toronto
</td>
<td>Garden District, Ryerson
</td></tr>
<tr>
<td>M6B
</td>
<td>North York
</td>
<td>Glencairn
</td></tr>
<tr>
<td>M7B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9B
</td>
<td>Etobicoke
</td>
<td>West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
</td></tr>
<tr>
<td>M1C
</td>
<td>Scarborough
</td>
<td>Rouge Hill, Port Union, Highland Creek
</td></tr>
<tr>
<td>M2C
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3C
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4C
</td>
<td>East York
</td>
<td>Woodbine Heights
</td></tr>
<tr>
<td>M5C
</td>
<td>Downtown Toronto
</td>
<td>St. James Town
</td></tr>
<tr>
<td>M6C
</td>
<td>York
</td>
<td>Humewood-Cedarvale
</td></tr>
<tr>
<td>M7C
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8C
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9C
</td>
<td>Etobicoke
</td>
<td>Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
</td></tr>
<tr>
<td>M1E
</td>
<td>Scarborough
</td>
<td>Guildwood, Morningside, West Hill
</td></tr>
<tr>
<td>M2E
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3E
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4E
</td>
<td>East Toronto
</td>
<td>The Beaches
</td></tr>
<tr>
<td>M5E
</td>
<td>Downtown Toronto
</td>
<td>Berczy Park
</td></tr>
<tr>
<td>M6E
</td>
<td>York
</td>
<td>Caledonia-Fairbanks
</td></tr>
<tr>
<td>M7E
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8E
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9E
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M1G
</td>
<td>Scarborough
</td>
<td>Woburn
</td></tr>
<tr>
<td>M2G
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3G
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4G
</td>
<td>East York
</td>
<td>Leaside
</td></tr>
<tr>
<td>M5G
</td>
<td>Downtown Toronto
</td>
<td>Central Bay Street
</td></tr>
<tr>
<td>M6G
</td>
<td>Downtown Toronto
</td>
<td>Christie
</td></tr>
<tr>
<td>M7G
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8G
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9G
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M1H
</td>
<td>Scarborough
</td>
<td>Cedarbrae
</td></tr>
<tr>
<td>M2H
</td>
<td>North York
</td>
<td>Hillcrest Village
</td></tr>
<tr>
<td>M3H
</td>
<td>North York
</td>
<td>Bathurst Manor, Wilson Heights, Downsview North
</td></tr>
<tr>
<td>M4H
</td>
<td>East York
</td>
<td>Thorncliffe Park
</td></tr>
<tr>
<td>M5H
</td>
<td>Downtown Toronto
</td>
<td>Richmond, Adelaide, King
</td></tr>
<tr>
<td>M6H
</td>
<td>West Toronto
</td>
<td>Dufferin, Dovercourt Village
</td></tr>
<tr>
<td>M7H
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8H
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9H
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M1J
</td>
<td>Scarborough
</td>
<td>Scarborough Village
</td></tr>
<tr>
<td>M2J
</td>
<td>North York
</td>
<td>Fairview, Henry Farm, Oriole
</td></tr>
<tr>
<td>M3J
</td>
<td>North York
</td>
<td>Northwood Park, York University
</td></tr>
<tr>
<td>M4J
</td>
<td>East York
</td>
<td>East Toronto, Broadview North (Old East York)
</td></tr>
<tr>
<td>M5J
</td>
<td>Downtown Toronto
</td>
<td>Harbourfront East, Union Station, Toronto Islands
</td></tr>
<tr>
<td>M6J
</td>
<td>West Toronto
</td>
<td>Little Portugal, Trinity
</td></tr>
<tr>
<td>M7J
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8J
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9J
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M1K
</td>
<td>Scarborough
</td>
<td>Kennedy Park, Ionview, East Birchmount Park
</td></tr>
<tr>
<td>M2K
</td>
<td>North York
</td>
<td>Bayview Village
</td></tr>
<tr>
<td>M3K
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4K
</td>
<td>East Toronto
</td>
<td>The Danforth West, Riverdale
</td></tr>
<tr>
<td>M5K
</td>
<td>Downtown Toronto
</td>
<td>Toronto Dominion Centre, Design Exchange
</td></tr>
<tr>
<td>M6K
</td>
<td>West Toronto
</td>
<td>Brockton, Parkdale Village, Exhibition Place
</td></tr>
<tr>
<td>M7K
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8K
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9K
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M1L
</td>
<td>Scarborough
</td>
<td>Golden Mile, Clairlea, Oakridge
</td></tr>
<tr>
<td>M2L
</td>
<td>North York
</td>
<td>York Mills, Silver Hills
</td></tr>
<tr>
<td>M3L
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4L
</td>
<td>East Toronto
</td>
<td>India Bazaar, The Beaches West
</td></tr>
<tr>
<td>M5L
</td>
<td>Downtown Toronto
</td>
<td>Commerce Court, Victoria Hotel
</td></tr>
<tr>
<td>M6L
</td>
<td>North York
</td>
<td>North Park, Maple Leaf Park, Upwood Park
</td></tr>
<tr>
<td>M7L
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8L
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9L
</td>
<td>North York
</td>
<td>Humber Summit
</td></tr>
<tr>
<td>M1M
</td>
<td>Scarborough
</td>
<td>Cliffside, Cliffcrest, Scarborough Village West
</td></tr>
<tr>
<td>M2M
</td>
<td>North York
</td>
<td>Willowdale, Newtonbrook
</td></tr>
<tr>
<td>M3M
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4M
</td>
<td>East Toronto
</td>
<td>Studio District
</td></tr>
<tr>
<td>M5M
</td>
<td>North York
</td>
<td>Bedford Park, Lawrence Manor East
</td></tr>
<tr>
<td>M6M
</td>
<td>York
</td>
<td>Del Ray, Mount Dennis, Keelsdale and Silverthorn
</td></tr>
<tr>
<td>M7M
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8M
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9M
</td>
<td>North York
</td>
<td>Humberlea, Emery
</td></tr>
<tr>
<td>M1N
</td>
<td>Scarborough
</td>
<td>Birch Cliff, Cliffside West
</td></tr>
<tr>
<td>M2N
</td>
<td>North York
</td>
<td>Willowdale, Willowdale East
</td></tr>
<tr>
<td>M3N
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4N
</td>
<td>Central Toronto
</td>
<td>Lawrence Park
</td></tr>
<tr>
<td>M5N
</td>
<td>Central Toronto
</td>
<td>Roselawn
</td></tr>
<tr>
<td>M6N
</td>
<td>York
</td>
<td>Runnymede, The Junction North
</td></tr>
<tr>
<td>M7N
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8N
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9N
</td>
<td>York
</td>
<td>Weston
</td></tr>
<tr>
<td>M1P
</td>
<td>Scarborough
</td>
<td>Dorset Park, Wexford Heights, Scarborough Town Centre
</td></tr>
<tr>
<td>M2P
</td>
<td>North York
</td>
<td>York Mills West
</td></tr>
<tr>
<td>M3P
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4P
</td>
<td>Central Toronto
</td>
<td>Davisville North
</td></tr>
<tr>
<td>M5P
</td>
<td>Central Toronto
</td>
<td>Forest Hill North &amp; West, Forest Hill Road Park
</td></tr>
<tr>
<td>M6P
</td>
<td>West Toronto
</td>
<td>High Park, The Junction South
</td></tr>
<tr>
<td>M7P
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8P
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9P
</td>
<td>Etobicoke
</td>
<td>Westmount
</td></tr>
<tr>
<td>M1R
</td>
<td>Scarborough
</td>
<td>Wexford, Maryvale
</td></tr>
<tr>
<td>M2R
</td>
<td>North York
</td>
<td>Willowdale, Willowdale West
</td></tr>
<tr>
<td>M3R
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4R
</td>
<td>Central Toronto
</td>
<td>North Toronto West,  Lawrence Park
</td></tr>
<tr>
<td>M5R
</td>
<td>Central Toronto
</td>
<td>The Annex, North Midtown, Yorkville
</td></tr>
<tr>
<td>M6R
</td>
<td>West Toronto
</td>
<td>Parkdale, Roncesvalles
</td></tr>
<tr>
<td>M7R
</td>
<td>Mississauga
</td>
<td>Canada Post Gateway Processing Centre
</td></tr>
<tr>
<td>M8R
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9R
</td>
<td>Etobicoke
</td>
<td>Kingsview Village, St. Phillips, Martin Grove Gardens, Richview Gardens
</td></tr>
<tr>
<td>M1S
</td>
<td>Scarborough
</td>
<td>Agincourt
</td></tr>
<tr>
<td>M2S
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3S
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4S
</td>
<td>Central Toronto
</td>
<td>Davisville
</td></tr>
<tr>
<td>M5S
</td>
<td>Downtown Toronto
</td>
<td>University of Toronto, Harbord
</td></tr>
<tr>
<td>M6S
</td>
<td>West Toronto
</td>
<td>Runnymede, Swansea
</td></tr>
<tr>
<td>M7S
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8S
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9S
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M1T
</td>
<td>Scarborough
</td>
<td>Clarks Corners, Tam O'Shanter, Sullivan
</td></tr>
<tr>
<td>M2T
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3T
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4T
</td>
<td>Central Toronto
</td>
<td>Moore Park, Summerhill East
</td></tr>
<tr>
<td>M5T
</td>
<td>Downtown Toronto
</td>
<td>Kensington Market, Chinatown, Grange Park
</td></tr>
<tr>
<td>M6T
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M7T
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8T
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9T
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M1V
</td>
<td>Scarborough
</td>
<td>Milliken, Agincourt North, Steeles East, L'Amoreaux East
</td></tr>
<tr>
<td>M2V
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3V
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4V
</td>
<td>Central Toronto
</td>
<td>Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
</td></tr>
<tr>
<td>M5V
</td>
<td>Downtown Toronto
</td>
<td>CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
</td></tr>
<tr>
<td>M6V
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M7V
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8V
</td>
<td>Etobicoke
</td>
<td>New Toronto, Mimico South, Humber Bay Shores
</td></tr>
<tr>
<td>M9V
</td>
<td>Etobicoke
</td>
<td>South Steeles, Silverstone, Humbergate, Jamestown, Mount Olive, Beaumond Heights, Thistletown, Albion Gardens
</td></tr>
<tr>
<td>M1W
</td>
<td>Scarborough
</td>
<td>Steeles West, L'Amoreaux West
</td></tr>
<tr>
<td>M2W
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3W
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4W
</td>
<td>Downtown Toronto
</td>
<td>Rosedale
</td></tr>
<tr>
<td>M5W
</td>
<td>Downtown Toronto
</td>
<td>Stn A PO Boxes
</td></tr>
<tr>
<td>M6W
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M7W
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8W
</td>
<td>Etobicoke
</td>
<td>Alderwood, Long Branch
</td></tr>
<tr>
<td>M9W
</td>
<td>Etobicoke
</td>
<td>Northwest, West Humber - Clairville
</td></tr>
<tr>
<td>M1X
</td>
<td>Scarborough
</td>
<td>Upper Rouge
</td></tr>
<tr>
<td>M2X
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3X
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4X
</td>
<td>Downtown Toronto
</td>
<td>St. James Town, Cabbagetown
</td></tr>
<tr>
<td>M5X
</td>
<td>Downtown Toronto
</td>
<td>First Canadian Place, Underground city
</td></tr>
<tr>
<td>M6X
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M7X
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8X
</td>
<td>Etobicoke
</td>
<td>The Kingsway, Montgomery Road, Old Mill North
</td></tr>
<tr>
<td>M9X
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M1Y
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2Y
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3Y
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4Y
</td>
<td>Downtown Toronto
</td>
<td>Church and Wellesley
</td></tr>
<tr>
<td>M5Y
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M6Y
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M7Y
</td>
<td>East Toronto
</td>
<td>Business reply mail Processing Centre, South Central Letter Processing Plant Toronto
</td></tr>
<tr>
<td>M8Y
</td>
<td>Etobicoke
</td>
<td>Old Mill South, King's Mill Park, Sunnylea, Humber Bay, Mimico NE, The Queensway East, Royal York South East, Kingsway Park South East
</td></tr>
<tr>
<td>M9Y
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M1Z
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2Z
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3Z
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M4Z
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M5Z
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M6Z
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M7Z
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M8Z
</td>
<td>Etobicoke
</td>
<td>Mimico NW, The Queensway West, South of Bloor, Kingsway Park South West, Royal York South West
</td></tr>
<tr>
<td>M9Z
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr></tbody></table> """



In [72]:
import pandas as pd

table2 = pd.read_html(table)
df = pd.DataFrame(table2[0]) 
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [74]:
df = pd.DataFrame(table2[0]) 
df



Postal Code           Borough               Neighborhood
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

In [75]:

df = df[df['Borough'] != "Not assigned"]
df.shape


(103, 3)

More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

In [83]:
#df=df.groupby(by=['Postal Code']).sum().reset_index()
#df[df["Postal Code"]=="M5A"]
pd.set_option('display.max_rows', None)
df.groupby(by=['Postal Code']).count()


Borough  Neighborhood
Postal Code                       
M1B                1             1
M1C                1             1
M1E                1             1
M1G                1             1
M1H                1             1
M1J                1             1
M1K                1             1
M1L                1             1
M1M                1             1
M1N                1             1
M1P                1             1
M1R                1             1
M1S                1             1
M1T                1             1
M1V                1             1
M1W                1             1
M1X                1             1
M2H                1             1
M2J                1             1
M2K                1             1
M2L                1             1
M2M                1             1
M2N                1             1
M2P                1             1
M2R                1             1
M3A                1             1
M3B                1             1
M3C                1             1
M3H                1             1
M3J                1             1
M3K                1             1
M3L                1             1
M3M                1             1
M3N                1             1
M4A                1             1
M4B                1             1
M4C                1             1
M4E                1             1
M4G                1             1
M4H                1             1
M4J                1             1
M4K                1             1
M4L                1             1
M4M                1             1
M4N                1             1
M4P                1             1
M4R                1             1
M4S                1             1
M4T                1             1
M4V                1             1
M4W                1             1
M4X                1             1
M4Y                1             1
M5A                1             1
M5B                1             1
M5C                1             1
M5E                1             1
M5G                1             1
M5H                1             1
M5J                1             1
M5K                1             1
M5L                1             1
M5M                1             1
M5N                1             1
M5P                1             1
M5R                1             1
M5S                1             1
M5T                1             1
M5V                1             1
M5W                1             1
M5X                1             1
M6A                1             1
M6B                1             1
M6C                1             1
M6E                1             1
M6G                1             1
M6H                1             1
M6J                1             1
M6K                1             1
M6L                1             1
M6M                1             1
M6N                1             1
M6P                1             1
M6R                1             1
M6S                1             1
M7A                1             1
M7R                1             1
M7Y                1             1
M8V                1             1
M8W                1             1
M8X                1             1
M8Y                1             1
M8Z                1             1
M9A                1             1
M9B                1             1
M9C                1             1
M9L                1             1
M9M                1             1
M9N                1             1
M9P                1             1
M9R                1             1
M9V                1             1
M9W                1             1

Maybe I am missing something but it seems to me that I have no duplicate in this dataset

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [88]:

df[df["Neighborhood"]=="Not assigned"]

df["Neighborhood"].unique


<bound method Series.unique of 2                                              Parkwoods
3                                       Victoria Village
4                              Regent Park, Harbourfront
5                       Lawrence Manor, Lawrence Heights
6            Queen's Park, Ontario Provincial Government
8                Islington Avenue, Humber Valley Village
9                                         Malvern, Rouge
11                                             Don Mills
12                       Parkview Hill, Woodbine Gardens
13                              Garden District, Ryerson
14                                             Glencairn
17     West Deane Park, Princess Gardens, Martin Grov...
18                Rouge Hill, Port Union, Highland Creek
20                                             Don Mills
21                                      Woodbine Heights
22                                        St. James Town
23                                    Humewood-Cedarvale


It also seems there are no "Not assigned" neighborhood in the data set


In [90]:
df.shape

(103, 3)

## PART II

I do not manage to get the latitude and longitde from the geo package, so I will just improt the data set for the CSV file.

In [97]:
df2=pd.read_csv('Geospatial_Coordinates.csv')

In [99]:
df3=pd.merge(df,
df2,
how='left', on="Postal Code")

In [101]:
df3.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [109]:
df4=df3[df3['Borough'].str.contains("Toronto")]
df4


Postal Code           Borough  \
2           M5A  Downtown Toronto   
4           M7A  Downtown Toronto   
9           M5B  Downtown Toronto   
15          M5C  Downtown Toronto   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
54          M4M      East Toronto   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   

                                          Neighborhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68     Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                   North Toronto West, Lawrence Park  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                                  Runnymede, Swansea  43.651571 -79.484450  
83                         Moore Park, Summerhill East  43.689574 -79.383160  
84           Kensington Market, Chinatown, Grange Park  43.653206 -79.4000

Getting the lat and long for Toronto

In [103]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [104]:
address = 'Toronto, CANADA'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))




The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Now I just explore the Toronto neighborhoods on a map

In [114]:
# create map of Manhattan using latitude and longitude values
map_tor = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df4['Latitude'], df4['Longitude'],df4["Neighborhood"]):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tor)  
    
map_tor

In [116]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [117]:
toronto_venues = getNearbyVenues(names=df4['Neighborhood'],
                                   latitudes=df4['Latitude'],
                                   longitudes=df4['Longitude']
                                  )



Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [118]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 195 uniques categories.


We have 195 different types of venues in the Toronto neighborhoods

I will now prepare the dataset for clustering.
First I do some one hot encoding.

In [132]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 



toronto_onehot.head()



Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       1     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Basketball Stadium  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                   0         0           0                   0       0   
1                   0         0           0                   0       0   
2                   0         0           0                   0       0   
3                   0         0           0                   0       0   
4                   0         0           0                   0       0   

   Boat or Ferry  Bookstore  Breakfast Spot  Brewery  Bubble Tea Shop  \
0              0          0               0        0                0   
1              0          0               0        0                0   
2              0          0               0        0                0   
3              0          0               0        0                0   
4              0          0               0        0                0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Candy Store  Caribbean Restaurant  Cheese Shop  \
0                          0            0                     0            0   
1                          0            0                     0            0   
2                          0            0                     0            0   
3                          0            0                     0            0   
4                          0            0                     0            0   

   Chinese Restaurant  Chocolate Shop  Church  Climbing Gym  Clothing Store  \
0                   0               0       0             0               0   
1                   0               0       0             0               0   
2                   0               0       0             0               0   
3                   0               0       0             0               0   
4     

In [134]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Airport  \
0                                         Berczy Park   0.0000   
1        Brockton, Parkdale Village, Exhibition Place   0.0000   
2   Business reply mail Processing Centre, South C...   0.0000   
3   CN Tower, King and Spadina, Railway Lands, Har...   0.0625   
4                                  Central Bay Street   0.0000   
5                                            Christie   0.0000   
6                                Church and Wellesley   0.0000   
7                      Commerce Court, Victoria Hotel   0.0000   
8                                          Davisville   0.0000   
9                                    Davisville North   0.0000   
10                       Dufferin, Dovercourt Village   0.0000   
11             First Canadian Place, Underground city   0.0000   
12    Forest Hill North & West, Forest Hill Road Park   0.0000   
13                           Garden District, Ryerson   0.0000   
14  Harbourfront East, Union Station, Toronto Islands   0.0000   
15                      High Park, The Junction South   0.0000   
16                     India Bazaar, The Beaches West   0.0000   
17          Kensington Market, Chinatown, Grange Park   0.0000   
18                                      Lawrence Park   0.0000   
19                           Little Portugal, Trinity   0.0000   
20                        Moore Park, Summerhill East   0.0000   
21                  North Toronto West, Lawrence Park   0.0000   
22                             Parkdale, Roncesvalles   0.0000   
23        Queen's Park, Ontario Provincial Government   0.0000   
24                          Regent Park, Harbourfront   0.0000   
25                           Richmond, Adelaide, King   0.0000   
26                                           Rosedale   0.0000   
27                                           Roselawn   0.0000   
28                                 Runnymede, Swansea   0.0000   
29                                     St. James Town   0.0000   
30                        St. James Town, Cabbagetown   0.0000   
31                                     Stn A PO Boxes   0.0000   
32                                    Studio District   0.0000   
33  Summerhill West, Rathnelly, South Hill, Forest...   0.0000   
34                The Annex, North Midtown, Yorkville   0.0000   
35                                        The Beaches   0.0000   
36                       The Danforth West, Riverdale   0.0000   
37           Toronto Dominion Centre, Design Exchange   0.0000   
38                     University of Toronto, Harbord   0.0000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0               0.0000        0.0000          0.0000           0.0000   
1               0.0000        0.0000          0.0000           0.0000   
2               0.0000        0.0000          0.0000           0.0000   
3               0.0625        0.0625          0.0625           0.1875   
4               0.0000        0.0000          0.0000           0.0000   
5               0.0000        0.0000          0.0000           0.0000   
6               0.0000        0.0000          0.0000           0.0000   
7               0.0000        0.0000          0.0000           0.0000   
8               0.0000        0.0000          0.0000           0.0000   
9               0.0000        0.0000          0.0000           0.0000   
10              0.0000        0.0000          0.0000           0.0000   
11              0.0000        0.0000          0.0000           0.0000   
12              0.0000        0.0000          0.0000           0.0000   
13              0.0000        0.0000          0.0000           0.0000   
14              0.0000        0.0000          0.0000           0.0000   
15              0.0000        0.0000          0.0000           0.0000   
16              0.0000        0.0000          0.0000           0.0000   
17              0.0000        0.0000          0.0000           0.0000   
18           

I will create a dataframe with the 15 top venues in the neighborhood

In [136]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


num_top_venues = 15

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park          Cocktail Bar   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Coffee Shop              Beer Bar    Seafood Restaurant   
1  Performing Arts Venue        Breakfast Spot           Coffee Shop   
2                    Spa  Gym / Fitness Center  Fast Food Restaurant   
3       Airport Terminal               Airport       Harbor / Marina   
4    Japanese Restaurant                  Café           Yoga Studio   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Restaurant                Bakery    Basketball Stadium   
1                      Bakery                   Bar         Burrito Place   
2              Farmers Market    Light Rail Station            Comic Shop   
3                 Coffee Shop                 Plane   Rental Car Location   
4  Modern European Restaurant        Sandwich Place       Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  \
0        Breakfast Spot             Jazz Club                       Café   
1            Restaurant          Climbing Gym                Yoga Studio   
2                  Park           Pizza Place           Recording Studio   
3      Sculpture Garden                   Bar              Boat or Ferry   
4      Ramen Restaurant    Chinese Restaurant  Middle Eastern Restaurant   

  11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
0            Fish Market           Liquor Store               Creperie   
1              Nightclub      Convenience Store     Italian Restaurant   
2             Restaurant          Burrito Place                Brewery   
3     Airport Food Court         Airport Lounge           Airport Gate   
4             Steakhouse             Comic Shop           Dessert Shop   

  14th Most Common Venue 15th Most Common Venue  
0                   Park           Concert Hall  
1           Intersection                    Gym  
2             Skate Park             Smoke Shop  
3                 Garden          Deli / Bodega  
4     Italian Restaurant                  Hotel

Now I start the clustering

In [153]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 2, 0, 0, 2, 2, 0], dtype=int32)

In [157]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Clusters Labels', kmeans.labels_)

toronto_merged = df4

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')



In [158]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Clusters Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Let's see if we can find out what differentiate the clusters

In [165]:
toronto_merged.loc[toronto_merged['Clusters Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Clusters Labels 1st Most Common Venue  \
15   Downtown Toronto                0             Gastropub   
19       East Toronto                0           Pizza Place   
20   Downtown Toronto                0          Cocktail Bar   
25   Downtown Toronto                0         Grocery Store   
31       West Toronto                0                Bakery   
36   Downtown Toronto                0                  Park   
37       West Toronto                0                   Bar   
41       East Toronto                0      Greek Restaurant   
43       West Toronto                0                  Café   
47       East Toronto                0           Pizza Place   
67    Central Toronto                0     Food & Drink Shop   
69       West Toronto                0    Mexican Restaurant   
75       West Toronto                0             Gift Shop   
80   Downtown Toronto                0                  Café   
84   Downtown Toronto                0                  Café   
87   Downtown Toronto                0       Airport Service   
92   Downtown Toronto                0          Cocktail Bar   
96   Downtown Toronto                0                Bakery   
99   Downtown Toronto                0          Burger Joint   
100      East Toronto                0           Yoga Studio   

     2nd Most Common Venue  3rd Most Common Venue  \
15                    Café            Coffee Shop   
19                   Trail                    Pub   
20             Coffee Shop               Beer Bar   
25                    Café                   Park   
31                Pharmacy                Brewery   
36                   Plaza                   Café   
37        Asian Restaurant            Men's Store   
41      Italian Restaurant         Ice Cream Shop   
43   Performing Arts Venue         Breakfast Spot   
47    Fast Food Restaurant                    Pub   
67                     Gym                   Park   
69                    Café        Thai Restaurant   
75          Breakfast Spot       Cuban Restaurant   
80      Italian Restaurant             Restaurant   
84                  Bakery  Vietnamese Restaurant   
87        Airport Terminal                Airport   
92                    Café               Creperie   
96             Coffee Shop             Restaurant   
99                 Gay Bar                   Park   
100                    Spa   Gym / Fitness Center   

             4th Most Common Venue 5th Most Common Venue  \
15                            Park                 Hotel   
19               Health Food Store      Asian Restaurant   
20              Seafood Restaurant            Restaurant   
25                     Candy Store                 Diner   
31       Middle Eastern Restaurant      Recording Studio   
36                           Hotel       Bubble Tea Shop   
37                     Yoga Studio   Japanese Restaurant   
41                     Yoga Studio                   Pub   
43                     Coffee Shop                Bakery   
47                         Brewery            Restaurant   
67                  Sandwich Place      Department Store   
69              Italian Restaurant           Flea Market   
75                         Dog Run                   Bar   
80                       Bookstore   Japanese Restaurant   
84   Vegetarian / Vegan Restaurant    Mexican Restaurant   
87                 Harbor / Marina           Coffee Shop   
92                        Beer Bar            Food Truck   
96              Italian Restaurant                  Café   
99                      Smoke Shop  Ethiopian Restaurant   
100           Fast Food Restaurant        Farmers Market   

    6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
15         Ice Cream Shop    Italian Restaurant        Japanese Restaurant   
19            Yoga Studio      Cuban Restaurant                 Donut Shop   
20                 Bakery    Basketball Stadium             

I find it hard to say what is defining cluster 0, maybe fast foods or cafe ?

In [166]:
toronto_merged.loc[toronto_merged['Clusters Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Clusters Labels 1st Most Common Venue  \
62  Central Toronto                1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62           Yoga Studio          Dance Studio   Dumpling Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62            Donut Shop               Dog Run   Distribution Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
62        Discount Store                 Diner           Dessert Shop   

   11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
62       Department Store          Deli / Bodega       Cuban Restaurant   

   14th Most Common Venue 15th Most Common Venue  
62      Electronics Store               Creperie

Cluster one is the only one with a gardern

In [170]:
toronto_merged.loc[toronto_merged['Clusters Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]



Borough  Clusters Labels 1st Most Common Venue  \
2   Downtown Toronto                2           Coffee Shop   
4   Downtown Toronto                2           Coffee Shop   
9   Downtown Toronto                2                  Café   
24  Downtown Toronto                2           Coffee Shop   
30  Downtown Toronto                2           Coffee Shop   
42  Downtown Toronto                2           Coffee Shop   
48  Downtown Toronto                2                  Café   
54      East Toronto                2                  Café   
73   Central Toronto                2        Clothing Store   
74   Central Toronto                2                  Café   
79   Central Toronto                2          Dessert Shop   
81      West Toronto                2      Sushi Restaurant   
86   Central Toronto                2                   Pub   
97  Downtown Toronto                2                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                   Park        Breakfast Spot                  Café   
4       Sushi Restaurant    Mexican Restaurant              Beer Bar   
9            Coffee Shop        Clothing Store               Theater   
24   Japanese Restaurant                  Café           Yoga Studio   
30                  Café           Pizza Place   Monument / Landmark   
42                  Café            Restaurant                Bakery   
48           Coffee Shop            Restaurant            Steakhouse   
54           Coffee Shop                Bakery           Yoga Studio   
73           Coffee Shop           Yoga Studio    Mexican Restaurant   
74        Sandwich Place           Coffee Shop          Liquor Store   
79    Italian Restaurant                   Gym           Pizza Place   
81                  Café           Coffee Shop    Italian Restaurant   
86           Coffee Shop            Restaurant      Sushi Restaurant   
97           Coffee Shop            Restaurant    Seafood Restaurant   

         5th Most Common Venue      6th Most Common Venue  \
2                       Bakery                    Theater   
4                Smoothie Shop             Sandwich Place   
9             Ramen Restaurant          Electronics Store   
24  Modern European Restaurant             Sandwich Place   
30                      Lounge                 Restaurant   
42                      Museum                Pizza Place   
48                   Bookstore             Ice Cream Shop   
54            Stationery Store  Middle Eastern Restaurant   
73          Seafood Restaurant                        Spa   
74                   BBQ Joint             History Museum   
79                        Café             Sandwich Place   
81                 Pizza Place                        Pub   
86                  Bagel Shop                 Sports Bar   
97                    Tea Room                 Steakhouse   

   7th Most Common Venue 8th Most Common Venue      9th Most Common Venue  \
2                    Pub         Historic Site               Dessert Shop   
4          Burrito Place                  Café                       Park   
9                  Hotel                 Diner             Sandwich Place   
24       Bubble Tea Shop      Ramen Restaurant         Chinese Restaurant   
30    Seafood Restaurant                 Hotel                 Smoke Shop   
42   Japanese Restaurant              Beer Bar                      Hotel   
48         Deli / Bodega                Bakery                        Pub   
54                  Park             Pet Store            Coworking Space   
73    Salon / Barbershop   Sporting Goods Shop                 Restaurant   
74     Indian Restaurant                   Pub  Middle Eastern Restaurant   
79      Sushi Restaurant           Coffee Shop           Toy / Game Store   
81    Falafel Restaurant                 Diner                  Bookstore   
86                  Bank          Liquor Store                Pizza P

Cluster number 2 is the coffee shop cluster

In [172]:
toronto_merged.loc[toronto_merged['Clusters Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Clusters Labels 1st Most Common Venue  \
61   Central Toronto                3                  Park   
68   Central Toronto                3         Jewelry Store   
91  Downtown Toronto                3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
61           Swim School              Bus Line           Yoga Studio   
68                 Trail      Sushi Restaurant                  Park   
91            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
61          Dance Studio            Donut Shop               Dog Run   
68         Garden Center      Cuban Restaurant            Donut Shop   
91              Creperie               Dog Run   Distribution Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
61   Distribution Center        Discount Store                  Diner   
68               Dog Run   Distribution Center         Discount Store   
91        Discount Store                 Diner           Dessert Shop   

   11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
61           Dessert Shop       Department Store          Deli / Bodega   
68                  Diner           Dessert Shop       Department Store   
91       Department Store          Deli / Bodega           Dance Studio   

   14th Most Common Venue       15th Most Common Venue  
61       Cuban Restaurant  Eastern European Restaurant  
68          Deli / Bodega                 Dance Studio  
91       Cuban Restaurant               Cosmetics Shop

Cluster 3 is the Park cluster

In [173]:
toronto_merged.loc[toronto_merged['Clusters Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Clusters Labels 1st Most Common Venue  \
83  Central Toronto                4                   Gym   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
83                 Trail           Yoga Studio          Dance Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
83   Dumpling Restaurant            Donut Shop               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
83   Distribution Center        Discount Store                  Diner   

   11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
83           Dessert Shop       Department Store          Deli / Bodega   

   14th Most Common Venue 15th Most Common Venue  
83               Creperie       Cuban Restaurant

Central Toronto is the only one in cluster 4 and it seems to be the sport cluster !